src/pidROS.cpp

```diff
@@ -0,0 +1,299 @@
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/*
+  Author: Melonee Wise
+  Contributors: Dave Coleman, Jonathan Bohren, Bob Holmberg, Wim Meeussen
+  Desc: Implements a standard proportional-integral-derivative controller
+*/
+
+#include <algorithm>
+#include <cmath>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "control_toolbox/pidROS.hpp"
+
+namespace control_toolbox
+{
+
+PidROS::PidROS(rclcpp::Node::SharedPtr & node, std::string topic_prefix)
+: node_(node)
+{
+  if (topic_prefix.back() != '.' && !topic_prefix.empty()) {
+    topic_prefix_ = topic_prefix + ".";
+  } else {
+    topic_prefix_ = topic_prefix;
+  }
+
+  state_pub_ = node->create_publisher<control_msgs::msg::PidState>(
+    topic_prefix + "/pid_state", rclcpp::SensorDataQoS());
+  rt_state_pub_.reset(
+    new realtime_tools::RealtimePublisher<control_msgs::msg::PidState>(state_pub_));
+}
+
+PidROS::~PidROS()
+{
+}
+
+bool PidROS::getBooleanParam(const std::string & param_name, bool & value)
+{
+  rclcpp::Parameter param;
+  if (node_->has_parameter(param_name)) {
+    node_->get_parameter(param_name, param);
+    value = param.as_bool();
+    return true;
+  } else {
+    return false;
+  }
+}
+
+bool PidROS::getDoubleParam(const std::string & param_name, double & value)
+{
+  rclcpp::Parameter param;
+  if (node_->has_parameter(param_name)) {
+    node_->get_parameter(param_name, param);
+    value = param.as_double();
+    RCLCPP_INFO_STREAM(
+      node_->get_logger(), "parameter '" << param_name << "' in node '" << node_->get_name() <<
+        "' value is " << value << std::endl);
+    return true;
+  } else {
+    RCLCPP_INFO_STREAM(
+      node_->get_logger(), "parameter '" << param_name << "' in node '" << node_->get_name() <<
+        "' does not exists" << std::endl);
+    return false;
+  }
+}
+
+bool PidROS::initPid()
+{
+  double p, i, d, i_min, i_max;
+  bool antiwindup = false;
+  bool all_param_available = true;
```

Comment:
the grammar nazi in my closet is whispering to me

```suggestion
  bool all_params_available = true;
```